### PR TITLE
Provide support for Picolibc

### PR DESCRIPTION
**Pull request by keith-packard**
[Picolibc](https://keithp.com/picolibc/) is a libc implementation designed for and tested on small embedded systems.

Features:
 * Smaller stdio implementation (originally from avr-libc)
 * Malloc-free printf/scanf
 * "Exact" float/string conversions based on the Ryu method.
 * ABI-compatible thread-local storage (use __thread in applications)
 * Tested under QEMU on a range of RISC-V and ARM targets.

Porting to additional architectures requires initializing the per-CPU thread pointer in the context
switch code. On ARM, this requires a toolchain with TLS support (GCC in Debian is built correctly).

A sample program linked for cortex-m0 shows the following memory usage:

```
void main(void)
{
    	printf("Hello World! %s %d\n", CONFIG_BOARD, 12);
}
```
| text	  | data | bss  | dec    | hex  | version |
| ---     | ---  | ---  | ---    | ---  | ---     |
| 16908   | 428  | 4917 | 22253  | 56ed | minimal |
| 15460   | 464  | 4924 | 20848  | 5170 | picolibc integer-only |
| 20272   | 464  | 4924 | 25660  | 643c | picolibc full version |
| 20862   | 548  | 4936 | 26346  | 66ea | newlib-nano integer-only |
| 50542   | 916  | 4936 | 56394  | dc4a | newlib-nano full version |
| 48170   | 2924 | 4984 | 56078  | db0e | newlib |

### CODE REVIEW

File: cmake/toolchain/gnuriscvemb/generic.cmake
Comment (mbolivar-nordic) [inline cmake/toolchain/gnuriscvemb/generic.cmake:4]: Since GNURISCVEMB_TOOLCHAIN_PATH is new, why bother adding environment variable support for setting its value in an already deprecated state? A plain CMake variable should do, no?

File: boards/riscv/qemu_riscv32/qemu_riscv32.yaml
Comment (mbolivar-nordic) [inline boards/riscv/qemu_riscv32/qemu_riscv32.yaml:9]: Worth mentioning from toolchain_3rd_party_x_compilers.rst? @tejlmand

**Review by mbolivar-nordic — COMMENTED** [review]
Looks cool. I have a couple of drive-by build system questions.

File: boards/riscv/qemu_riscv32/qemu_riscv32.yaml
Comment (keith-packard) [reply boards/riscv/qemu_riscv32/qemu_riscv32.yaml:9]: Thanks, I'll check there and see about adding documentation elsewhere as well.

File: cmake/toolchain/gnuriscvemb/generic.cmake
Comment (keith-packard) [reply cmake/toolchain/gnuriscvemb/generic.cmake:4]: I'll see if I can figure this out; I just followed the existing gnuarmemb path for this work.

Comment (mbolivar-nordic) [reply cmake/toolchain/gnuriscvemb/generic.cmake:4]: More a question than a request for change, since I recall some historical back and forth around what configuration variables fall back to the environment if not set as ordinary CMake variables. I defer to @tejlmand as build system maintainer for what the right thing to do is here.

File: boards/riscv/qemu_riscv32/qemu_riscv32.yaml
Comment (mbolivar-nordic) [reply boards/riscv/qemu_riscv32/qemu_riscv32.yaml:9]: doc/guides/c_library.rst might also be a good place to mention the new Kconfig option, since that touches on how to use newlib.

Comment (keith-packard) [reply boards/riscv/qemu_riscv32/qemu_riscv32.yaml:9]: Thanks for the pointers - reviewing these docs, I discovered that what I want is the CROSS_COMPILE environment variable instead of gnuriscvemb. I think I should just remove that patch from the series.

**Comment by stephanosio** [comment]
Welcome to Zephyr! Glad to see that a regular newlib contributor is helping out with the project :)

File: cmake/extensions.cmake
Comment (stephanosio) [inline cmake/extensions.cmake:1390]: > Some linker options are only valid in combinations, like the RISC-V
> -march -mabi values. Attempting to validate them one-at-a-time will
> fail, even though the combination of values is valid. This could be
> changed to validate the whole set of options provided at one time, which
> would work in the RISC-V case.

We should not remove linker option validation just because it breaks something for RISC-V.

Please consider cherry-picking the following commit into this PR:
https://github.com/zephyrproject-rtos/zephyr/pull/25475/commits/f6a67ecc2f13c7c1efa1c47c32897b1e08e269d5

cc @tejlmand

**Review by stephanosio — CHANGES_REQUESTED** [review]

**Comment by stephanosio** [comment]
@galak @keith-packard I suppose we could potentially integrate this into the Zephyr SDK/toolchains by adding `pico.specs`.

File: cmake/extensions.cmake
Comment (keith-packard) [reply cmake/extensions.cmake]: Thanks so much for fixing this in a very clean way. My kludge was definitely not suitable for merging, but did manage to make the picolibc build work. Picolibc ships with a .specs file that computes the correct library name based on the arch/abi values provided to the compiler, so when those weren't passed to the linker, the wrong C library was used causing the link to fail. With your fix, the arguments are correctly passed and linking now works! 

I've cherry-picked that one patch into this PR; I'm happy to wait for your PR to be merged and rebase if you like.

Comment (stephanosio) [reply cmake/extensions.cmake]: Good to hear that it works.

> I've cherry-picked that one patch into this PR; I'm happy to wait for your PR to be merged and rebase if you like.

[That PR](https://github.com/zephyrproject-rtos/zephyr/pull/25475) will take quite some time to get merged. I am okay with that commit being merged as part of this PR. I will make sure to rebase my PR after that happens.

**Comment by keith-packard** [comment]
> @galak I suppose we could potentially integrate this into the Zephyr SDK/toolchains by adding `pico.specs`.
>
> @keith-packard Do you have any plans to add something like `pico.specs` for GCC integration so that various C libraries can be selected (e.g. `libc.a` for full newlib and `libc_pico.a` for picolibc)? Also any plans to support [crosstool-ng](https://github.com/crosstool-ng/crosstool-ng)? The Zephyr toolchains are currently built using crosstool-ng.

The PR already depends on the picolibc.specs file which is shipped and installed with picolibc itself, so there's no need to include an additional specs file in zephyr.

If there are additional zephyr-specific compiler configuration bits needed, we could create a separate .specs file that referenced the existing picolibc one. The advantage of using the installed picolibc.specs file is that it has the picolibc install paths configured so that you don't need any library or include paths for the toolchain.

**Comment by keith-packard** [comment]
> Welcome to Zephyr! Glad to see that a regular newlib contributor is helping out with the project :)

Been meaning to drop by for a while now. Thanks for the warm welcome.

File: cmake/extensions.cmake
Comment (keith-packard) [reply cmake/extensions.cmake]: Sounds good. I've tested the SDK's arm-zepyr-eabi compiler and it has the necessary TLS support enabled.

**Comment by stephanosio** [comment]
> I think you should be able to use the existing `picolibc.specs` file if you wanted to deliver picolibc as part of the Zephyr SDK/toolchains. The picolibc configuration supports sysroot-based GCC installs, which I suspect is what the SDK is using. That means you would be able to place picolibc relative to GCC install directory and `picolibc.specs` would then be able to locate picolibc wherever the SDK was installed.

I didn't know `picolibc.specs` existed. I will look into it.

It is worth noting that we will still need to specify `-specs=picolibc.specs` to make it work with the toolchains that support multiple C library variants, as we do for the nano newlib.

https://github.com/zephyrproject-rtos/zephyr/blob/aded6a53347ee924abbe4c394e578724b4e61681/lib/libc/newlib/CMakeLists.txt#L42-L49

**Comment by keith-packard** [comment]
> It is worth noting that we will still need to specify `-specs=picolibc.specs` to make it work with the toolchains that support multiple C library variants, as we do for the nano newlib.

Well, thanks for getting me to check -- somehow the lib/libc/picolibc directory got lost from the previous version of the patch. I've added that back, and you can see that lib/libc/picolibc/CMakeLists.txt adds --specs=picolibc.specs to both zephyr_link_libraries and zephyr_compile_options. Picolibc is designed to be installed in parallel with other C libraries.

**Comment by andrewboie** [comment]
Thanks for doing this, in particular getting proper compiler-supported TLS has been on my to-do list for a while.

The "CONFIG_USERSPACE_LOCAL_DATA" option just carves out a bit of room in the stack buffer for a fixed-sized "TLS" struct which currently just has `errno`. I think with a little more work we can unify all of this. No need to do it in this PR.

Will set aside some time later this week to look at this patch in detail.

File: lib/libc/picolibc/libc-hooks.c
Comment (cfriedt) [inline lib/libc/picolibc/libc-hooks.c:291]: There are only a few tests that are failing and it looks like they may have something to do with time. I'm only speculating, but maybe this is the reason.

Otherwise great work. Nice libc too ;-)

Comment (keith-packard) [reply lib/libc/picolibc/libc-hooks.c:291]: Given that none of the tests are using picolibc, I'm not quite sure why any of them failed when this code was added...

**Comment by keith-packard** [comment]
> The picolibc configuration supports sysroot-based GCC installs, which I suspect is what the SDK is using. 

Turns out the SDK *isn't* using --with-sysroot, so picolibc can't be used easily. I've filed an issue with that:

https://github.com/zephyrproject-rtos/sdk-ng/issues/236

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for this.
Looks very promising, will try do some target testing later for further evaluation.

File: cmake/compiler/gcc/target_riscv.cmake
Comment (tejlmand) [reply cmake/compiler/gcc/target_riscv.cmake:40]: I see @stephanosio actually did the commit, so he might have an answer.

Comment (keith-packard) [reply cmake/compiler/gcc/target_riscv.cmake:40]: Yes, without the " marks, each option is checked independently, which fails because they must be added together.

File: lib/libc/Kconfig
Comment (keith-packard) [reply lib/libc/Kconfig:32]: Sorry for using the wrong term; I was using 'SDK' in a more general way. But, yes, picolibc could be added to the Zephyr SDK as well so that users can get tooling which is known to work well. Perhaps 'expected to be already installed for use with the toolchain'.

File: doc/getting_started/toolchain_other_x_compilers.rst
Comment (keith-packard) [reply doc/getting_started/toolchain_other_x_compilers.rst:32]: I think we should sort out how picolibc should be delivered so that these instructions can be more specific. Getting picolibc included in the standard Zephyr toolchain build would certainly make everything easier for users.

File: doc/getting_started/toolchain_other_x_compilers.rst
Comment (keith-packard) [reply doc/getting_started/toolchain_other_x_compilers.rst:40]: Right, it's only available in the development version of Ubuntu: https://bugs.launchpad.net/ubuntu/+source/picolibc
We could write some instructions for getting picolibc installed from source, or I could create pre-built binaries against the current ARM embedded toolchain as I do for the current RISC-V embedded toolchain:
https://github.com/keith-packard/picolibc/releases/tag/1.4.6

File: cmake/compiler/gcc/target_riscv.cmake
Comment (tejlmand) [reply cmake/compiler/gcc/target_riscv.cmake:40]: Yes, except for the fact that the linker option is currently an empty string `""`.
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/extensions.cmake#L1359
```
    zephyr_check_compiler_flag(C "" ${check})
```
so that is why i'm curious exactly what is being solved with this change ?

File: lib/libc/Kconfig
Comment (tejlmand) [reply lib/libc/Kconfig:32]: Maybe you could rephrase this help text ?

File: doc/getting_started/toolchain_other_x_compilers.rst
Comment (tejlmand) [reply doc/getting_started/toolchain_other_x_compilers.rst:32]: Agreed. We already have the section `Other Cross Compilers`.

How about placing this in an independent section `Other C Libraries`
```
Other C Libraries
#################

It is possible to install another C library.

For example, you might install the picolibc-arm-none-eabi package
   on Debian-based Linux systems:

   .. code-block:: console

      # On Debian or Ubuntu
      sudo apt-get install picolibc-arm-none-eabi

.....
```
and then of course update the text with a bit more detailed information or link, to assist users on how to obtain picolib.
Of course, if/when picolib becomes part of Zephyr SDK, then it can just be referenced to Zephyr SDK.

File: doc/getting_started/toolchain_other_x_compilers.rst
Comment (tejlmand) [reply doc/getting_started/toolchain_other_x_compilers.rst:40]: I think that until picolibc is part of the stable software repositories (debian/Ubuntu) or Zephyr SDK, then I think it is important to include a bit more detailed description on how to obtain pre-built packages or how to build from source.

Just curious, can picolibc be used with `armgcc` on windows ?
I see it can be built and used natively on posix systems (which windows is not), but I don't see if it can be built and used by a cross compiler on windows (such as armgcc).

File: doc/getting_started/toolchain_other_x_compilers.rst
Comment (keith-packard) [reply doc/getting_started/toolchain_other_x_compilers.rst:32]: I've filed a PR with crosstools-ng to add picolibc support there; does that seem like a reasonable way to get picolibc support into the SDK?

 https://github.com/crosstool-ng/crosstool-ng/pull/1373

File: doc/getting_started/toolchain_other_x_compilers.rst
Comment (keith-packard) [reply doc/getting_started/toolchain_other_x_compilers.rst:40]: I haven't ever tried using armgcc on windows to build picolibc, but I don't know of any reason it *wouldn't* work, aside from GCC-isms in the code. picolibc inherits a lot of newlib code, and that does end up depending on GCC. Probably worth a try at least.

File: cmake/compiler/gcc/target_riscv.cmake
Comment (stephanosio) [reply cmake/compiler/gcc/target_riscv.cmake:40]: @tejlmand This does solve an error. Reading https://github.com/zephyrproject-rtos/zephyr/commit/5fb5adb34f5672749607a0b652da3e5fff0d6638 would help understand what is going on here.

File: doc/getting_started/toolchain_other_x_compilers.rst
Comment (keith-packard) [reply doc/getting_started/toolchain_other_x_compilers.rst:40]: I got picolibc compiling using armgcc under Windows over the weekend; meson needed mingw and msys to get the necessary tooling. I'll have patches merged to picolibc for the next release.

File: cmake/compiler/gcc/target_riscv.cmake
Comment (tejlmand) [reply cmake/compiler/gcc/target_riscv.cmake:40]: @stephanosio thanks for the link.

And thanks for confirming that this does fix an error.

It might be that the CMake documentation is a bit imprecise on this part (wouldn't be the first time), cause the docs only refers to `CMAKE_REQUIRED_FLAGS` being used for compilation and not linking.

https://cmake.org/cmake/help/latest/module/CheckCCompilerFlag.html
which actually uses `check_c_source_compiles`
https://cmake.org/cmake/help/latest/module/CheckCSourceCompiles.html 
>CMAKE_REQUIRED_FLAGS
>
>    Additional flags to pass to the compiler. ...

However, it could be that as long as we are using gcc based compiler and linker, then it is enough to check the flag on the compiler itself.
And as you confirm this actually fixes an issue you have seen, then i'm fine with taking in this change.

But I have a small concern that the current testing scheme is not complete (another story).
As long as we are using gcc based toolchains for both compiling and linking we might not see any errors.

However, this may change when starting to work with more toolchains, e.g. clang, Metaware, IAR, armclang.
Especially with clang there can be a difference on checking the flags for `ld` or `ldd`, and testing a flag only on the compiler may not be sufficient.

**Comment by keith-packard** [comment]
Ok, pretty big update today.

 * Rebased to current head
 * Added tests
 * Add aarch64 (including ilp32)
 * Fixed cortex-r5 TLS bug
 * Updated docs, but still need work

I'm now building picolibc using the zephyr SDK and installing them there; I have scripts that automate that for picolibc which I can publish in that repository to help make things easier for users. I'll be doing a picolibc release with the fixes related to this work fairly soon, and those scripts can be included at that point.

File: cmake/compiler/gcc/target_riscv.cmake
Comment (tejlmand) [reply cmake/compiler/gcc/target_riscv.cmake:40]: @stephanosio we should find another approach for ensuring those flags are checked together.

Quoting flags like this in CMake may have unintended side-effects.
It often works directly on compile options, but if the properties are fetched for other purposes, either using a generator expression or by `get_property(...)` then there is a risk that the flag may be quoted on command line.

One example is `external_lib`:
```
... -fno-common -g "-mabi=ilp32 -march=rv32ima" -imacros ...
riscv64-zephyr-elf-gcc: error: unrecognized argument in option '-mabi=ilp32 -march=rv32ima'
```

File: CMakeLists.txt
Comment (keith-packard) [reply CMakeLists.txt]: Good idea. I've updated the patch to include this

File: cmake/compiler/gcc/target_riscv.cmake
Comment (keith-packard) [reply cmake/compiler/gcc/target_riscv.cmake:40]: How about allowing the combined option in TOOLCHAIN_LD_FLAGS and then splitting it into separate options once validated by the compiler:

```diff --git a/cmake/extensions.cmake b/cmake/extensions.cmake
index 94b597051e..98d12e4e4c 100644
--- a/cmake/extensions.cmake
+++ b/cmake/extensions.cmake
@@ -1359,7 +1359,10 @@ function(target_ld_options target scope)
     zephyr_check_compiler_flag(C "" ${check})
     set(CMAKE_REQUIRED_FLAGS ${SAVED_CMAKE_REQUIRED_FLAGS})
 
-    target_link_libraries_ifdef(${check} ${target} ${scope} ${option})
+    string(REPLACE " " ";" options ${option})
+    foreach(suboption IN LISTS options)
+      target_link_libraries_ifdef(${check} ${target} ${scope} ${suboption})
+    endforeach()
   endforeach()
 endfunction()
 
```

Comment (mbolivar-nordic) [reply cmake/compiler/gcc/target_riscv.cmake:40]: (Drive-by comment; I haven't been following the whole discussion.)

Sounds like this might be a job for `separate_arguments()`.

```
$ cat ~/foo.cmake
set(foo "-mabi=ilp32 -march=rv32ma --some-option=\"with spaces in the value\"")

separate_arguments(foo_as_list UNIX_COMMAND ${foo})

message(STATUS "foo_as_list=${foo_as_list}")
foreach(arg ${foo_as_list})
  message(STATUS "(${arg})")
endforeach()

$ cmake -P ~/foo.cmake
-- foo_as_list=-mabi=ilp32;-march=rv32ma;--some-option=with spaces in the value
-- (-mabi=ilp32)
-- (-march=rv32ma)
-- (--some-option=with spaces in the value)
```

Comment (keith-packard) [reply cmake/compiler/gcc/target_riscv.cmake:40]: thanks for the reference -- looks like that's a better option than my 'REPLACE' line.

I've pushed an updated picolibc branch with this change.

File: arch/arm/core/aarch64/switch.S
Comment (carlocaione) [inline arch/arm/core/aarch64/switch.S:72]: Can you use the same spacing and comment style of the rest of the file?

Comment (keith-packard) [reply arch/arm/core/aarch64/switch.S:72]: Yes, of course, I've pushed an update with the formatting adjusted to match as best I could tell.

**Comment by keith-packard** [comment]
I've updated this patch to sit on top of current master and require THREAD_LOCAL_STORAGE for architectures which support that. The changes are now far smaller as they don't include the TLS code.

File: CMakeLists.txt
Comment (tejlmand) [inline CMakeLists.txt:933]: How about:
```
  if(CONFIG_PICOLIBC OR CONFIG_NEWLIB_LIBC)
    set(LIBC_PARTITIONS -l libc.a z_libc_partition)
  endif()
```

File: lib/libc/CMakeLists.txt
Comment (tejlmand) [inline lib/libc/CMakeLists.txt:11]: Please make an `elseif()` in this case.

```suggestion
if(CONFIG_PICOLIBC)
  add_subdirectory(picolibc)
elseif(CONFIG_NEWLIB_LIBC)
  add_subdirectory(newlib)
else()
  add_subdirectory(minimal)
endif()
```

File: lib/libc/picolibc/libc-hooks.c
Comment (tejlmand) [inline lib/libc/picolibc/libc-hooks.c:118]: why a newline here ?

File: lib/libc/picolibc/libc-hooks.c
Comment (tejlmand) [inline lib/libc/picolibc/libc-hooks.c:128]: why a newline here ?

File: lib/libc/picolibc/CMakeLists.txt
Comment (tejlmand) [inline lib/libc/picolibc/CMakeLists.txt:20]: No changes requested, but with the fact that Zephyr is working towards more toolchains, especially LLVM, and thus  also the support of compiler-rt runtime library, I was wondering if picolib could be used in that case.

Also this makes me wonder if the use of picolib should be restricted, based on toolchain in use.
But that would probably need to go into a bigger work at later time.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Thanks for the updated work.
Looks good, only a couple of small suggestions.

File: lib/libc/picolibc/CMakeLists.txt
Comment (keith-packard) [reply lib/libc/picolibc/CMakeLists.txt:20]: picolibc built with llvm still requires libgcc as that has all of the helper functions required by the underlying ABI, such as the __aeabi functions for ARM. -lgcc would be automatically inserted into the link command by the compiler front-end (either clang or gcc) if zephyr wasn't using -nostdlib. zephyr could probably remove that flag and simply let the compiler front end insert both -lc and -lgcc.

**Pull request closed without merging**
